### PR TITLE
Setup layout for flag selection

### PR DESCRIPTION
**Pull request by MICHAELMUNAVU83**
I setup the layout for flags selection in the new registration flow at

```
"/beta?step=select_white_flags"
"/beta?step=select_red_flags"
"/beta?step=select_green_flags"

```
Screenshot 2024-12-17 at 11 56 46

Screenshot 2024-12-17 at 11 56 37

Screenshot 2024-12-17 at 11 56 29

### CODE REVIEW

File: lib/animina_web/components/beta_registration_components.ex [hunk 1]
```diff
@@ -84,6 +84,77 @@ defmodule AniminaWeb.BetaRegistrationComponents do
...
+      <button class="flex w-full justify-center rounded-md bg-indigo-600 px-3 py-1.5 text-sm font-semibold leading-6 text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 ">
+        <%= with_locale(@language, fn -> %>
+          <%= gettext("Save flags") %>
+        <% end) %>
+      </button>
+    </div>
+    """
+  end
+
+  defp get_default_button_colors(color) do
```
Comment (wintermeyer) [inline lib/animina_web/components/beta_registration_components.ex]: I don't like the coding style here. Either do something like:

```elixir
defp get_default_button_colors(color) do
  case color do
    :green -> "hover:bg-green-50 bg-green-100 focus-visible:outline-green-100 text-green-600"
    :red -> "hover:bg-red-50 bg-red-100 focus-visible:outline-red-100 text-red-600"
    _ -> "hover:bg-indigo-50 bg-indigo-100 focus-visible:outline-indigo-100 text-indigo-600"
  end
end
```

or (**even better**) something like:

```elixir
defp get_default_button_colors(:green), do: "hover:bg-green-50 bg-green-100 focus-visible:outline-green-100 text-green-600"
defp get_default_button_colors(:red), do: "hover:bg-red-50 bg-red-100 focus-visible:outline-red-100 text-red-600"
defp get_default_button_colors(_), do: "hover:bg-indigo-50 bg-indigo-100 focus-visible:outline-indigo-100 text-indigo-600"
```

File: lib/animina_web/live/beta_register.ex [hunk 2]
```diff
@@ -28,6 +31,84 @@ defmodule AniminaWeb.BetaRegisterLive do
     {:ok, socket}
   end
 
+  @impl true
+
+  def handle_params(%{"step" => "select_white_flags"}, _url, socket) do
```
Comment (wintermeyer) [inline lib/animina_web/live/beta_register.ex]: There is a lot to DRY and refactor here. green, red and white are more or less the same code.

```elixir
def handle_params(%{"step" => step}, _url, socket) do
  {color, page_title_str, title_str, info_text_str} = step_info(step)

  {:noreply, assign_flags(socket, color, page_title_str, title_str, info_text_str)}
end

defp step_info("select_white_flags") do
  {
    :white,
    "Select your own flags",
    "Choose Your Own Flags",
    "We use flags to match people. You can select red and green flags later. But first tell us something about yourself and select up to %{number_of_flags} flags that describe yourself. The ones selected first are the most important."
  }
end

defp step_info("select_green_flags") do
  {
    :green,
    "Select your green flags",
    "Choose Your Green Flags",
    "Choose up to %{number_of_flags} flags that you want your partner to have. The ones selected first are the most important."
  }
end

defp step_info("select_red_flags") do
  {
    :red,
    "Select your red flags",
    "Choose Your Red Flags",
    "Choose up to %{number_of_flags} flags that you don't want to have in a partner. The ones selected first are the most important."
  }
end

defp assign_flags(socket, color, page_title_str, title_str, info_text_str) do
  language = socket.assigns.language

  socket
  |> assign(:color, color)
  |> assign(:categories, fetch_categories())
  |> assign(:title, with_locale(language, fn -> gettext(title_str) end))
  |> assign(:info_text, with_locale(language, fn -> gettext(info_text_str, number_of_flags: @max_flags) end))
  |> assign(:page_title, with_locale(language, fn -> gettext(page_title_str) end))
  |> assign(:step, "select_flags")
end
```

Comment (wintermeyer) [reply lib/animina_web/live/beta_register.ex]: Please double check this. I just used ChatGPT to refactor it. I have never tested it.

**Review by wintermeyer — CHANGES_REQUESTED** [review]
@MICHAELMUNAVU83 Please start thinking a bit more about DRY and Elixir coding style. Use ChatGPT or other LLMs to help you to refactor code quickly. Less spaghetti code please.

File: lib/animina_web/live/beta_register.ex (hunk 2 above)
Comment (MICHAELMUNAVU83) [reply lib/animina_web/live/beta_register.ex]: Okay , I will take this into account in next PRs

File: lib/animina_web/components/beta_registration_components.ex (hunk 1 above)
Comment (wintermeyer) [reply lib/animina_web/components/beta_registration_components.ex]: One could argue that `default_button_colors/1` is a better naming than `get_default_button_colors/1`.

File: lib/animina_web/live/beta_register.ex [hunk 3]
```diff
@@ -28,6 +31,70 @@ defmodule AniminaWeb.BetaRegisterLive do
...
+      step_info(step, socket.assigns.language, @max_flags)
+
+    {:noreply, assign_flags(socket, color, page_title_str, title_str, info_text_str)}
+  end
+
+  def handle_params(_, _url, socket) do
+    {:noreply, socket |> assign(:step, "filter_potential_partners")}
+  end
+
+  defp step_info("select_white_flags", language, _) do
```
Comment (wintermeyer) [inline lib/animina_web/live/beta_register.ex:47]: Not a super big fan of this code. Example of an easier code:

```elixir
@steps_config %{
  "select_white_flags" => %{
    color: :white,
    page_title: "Select your own flags",
    title: "Choose Your Own Flags",
    info: "We use flags to match people. You can select red and green flags later. But first tell us something about yourself and select up to %{number_of_flags} flags that describe yourself. The ones selected first are the most important."
  },
  "select_green_flags" => %{
    color: :green,
    page_title: "Select your green flags",
    title: "Choose Your Green Flags",
    info: "Choose up to %{number_of_flags} flags that you want your partner to have. The ones selected first are the most important."
  },
  "select_red_flags" => %{
    color: :red,
    page_title: "Select your red flags",
    title: "Choose Your Red Flags",
    info: "Choose up to %{number_of_flags} flags that you don't want to have in a partner. The ones selected first are the most important."
  }
}

defp step_info(step, language, number_of_flags) do
  config = @steps_config[step]

  {
    config.color,
    with_locale(language, fn -> gettext(config.page_title) end),
    with_locale(language, fn -> gettext(config.title) end),
    with_locale(language, fn ->
      gettext(config.info, number_of_flags: number_of_flags)
    end)
  }
end
```

**Review by wintermeyer — APPROVED** [review]

**Pull request merged**